src/rules/jsxSortPropsRule.ts

```diff
@@ -0,0 +1,206 @@
+/**
+ * @license
+ * Copyright 2017 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Lint from "tslint";
+import {
+    isJsxAttribute,
+    isJsxOpeningLikeElement,
+    isJsxSpreadAttribute,
+} from "tsutils";
+import * as ts from "typescript";
+
+const OPTION_IGNORE_CASE = "ignore-case";
+
+const OPTION_ORDER_SPREAD = "spread";
+const OPTION_ORDER_SHORTHAND = "shorthand";
+const OPTION_ORDER_LONGHAND = "longhand";
+type ORDER_OPTION = "spread" | "shorthand" | "longhand";
+
+enum KEY_ORDER {
+  OPTION_ORDER_SPREAD,
+  OPTION_ORDER_SHORTHAND,
+  OPTION_ORDER_LONGHAND,
+}
+
+const BEHAVIOR_OPTIONS = {
+  enum: [
+    OPTION_IGNORE_CASE,
+  ],
+  type: "string",
+};
+
+const SORT_OPTIONS = {
+  items: {
+    enum: KEY_ORDER,
+    type: "string",
+  },
+  maxLength: 3,
+  type: "array",
+};
+
+interface IOptions {
+    ignoreCase: boolean;
+    order?: ORDER_OPTION[];
+}
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "jsx-sort-props",
+        description: "Requires props in JSX elements to be sorted alphabetically, and grouped as specified",
+        optionsDescription: Lint.Utils.dedent`
+            Two types of options are available:
+
+            Individual strings that control general functionality.
+            Possible settings are:
+
+            * \`"${OPTION_IGNORE_CASE}"\`: Ignore case when comparing props.
+
+            An array of strings that control ordering of props by type.
+            Possible settings are:
+
+            * \`"${OPTION_ORDER_SPREAD}"\`: Spread assignments \`{...props}\`.
+            * \`"${OPTION_ORDER_SHORTHAND}"\`: Shorthand properties \`{ isError, isValid }\`.
```

Comment:
shorthand conflicts with `jsx-boolean-value: true`, which we enable by default... https://github.com/palantir/tslint-react/blob/master/tslint-react.json#L5